File: Roboti/app/src/main/java/cz/muni/fi/umimecesky/roboti/activity/RaceActivity.java
package cz.muni.fi.umimecesky.roboti.activity;

import android.content.DialogInterface;
import android.os.Bundle;
import android.util.Log;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import cn.refactor.lib.colordialog.PromptDialog;
import cz.muni.fi.umimecesky.roboti.R;
import cz.muni.fi.umimecesky.roboti.pojo.Bot;
import cz.muni.fi.umimecesky.roboti.pojo.FillWord;
import cz.muni.fi.umimecesky.roboti.pojo.RaceConcept;
import cz.muni.fi.umimecesky.roboti.utils.CalculateDp;
import cz.muni.fi.umimecesky.roboti.utils.Constant;
import cz.muni.fi.umimecesky.roboti.utils.Global;
import cz.muni.fi.umimecesky.roboti.utils.MoveLogic;
import cz.muni.fi.umimecesky.roboti.utils.RobotDrawable;

import static cz.muni.fi.umimecesky.roboti.utils.Constant.RACE_NEW_WORD_DELAY;
import static cz.muni.fi.umimecesky.roboti.utils.Constant.RAW_HOPS_TO_WIN;

public class RaceActivity extends BaseAbstractActivity {

    private MoveLogic moveLogic;
    private Bot usersBot;

    private RaceConcept concept;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_race);
        concept = (RaceConcept) getIntent().getExtras().getSerializable(Constant.RACE_CONCEPT_EXTRA);

        ImageView usersRobot = (ImageView) findViewById(R.id.usersRobot);
        int hopsToWin = RAW_HOPS_TO_WIN + concept.getCurrentLevel();
        CalculateDp calculateDp = new CalculateDp(usersRobot, hopsToWin);
        calculateDp.setupFinishLine(findViewById(R.id.finishLine));
        Global.setCalculateDp(calculateDp);

        usersBot = new Bot(usersRobot, null); //TODO: change bot to robot, maybe add inheritance?

        setUpRobotViews();

        setWordText((TextView) findViewById(R.id.word));
        setVariant1((Button) findViewById(R.id.firstButton));
        setVariant2((Button) findViewById(R.id.secondButton));

        super.init();
        setNewRandomWord();

    }

    private void setUpRobotViews() {

        RobotDrawable robotDrawable = new RobotDrawable(this);
        ImageView usersRobot = (ImageView) findViewById(R.id.usersRobot);

        ImageView bot1 = (ImageView) findViewById(R.id.bot1);
        ImageView bot2 = (ImageView) findViewById(R.id.bot2);
        ImageView bot3 = (ImageView) findViewById(R.id.bot3);

        bot1.setImageDrawable(robotDrawable.removeRobotDrawable());
        bot2.setImageDrawable(robotDrawable.removeRobotDrawable());
        bot3.setImageDrawable(robotDrawable.removeRobotDrawable());

        moveLogic = new MoveLogic(this, concept, bot1, bot2, bot3);
    }

    protected void setNewRandomWord() {
        FillWord word = getWordCategoryHelper().getRandomCategoryWord(concept.getCategoryIDs());
        Log.d("random word", String.valueOf(word));
        setWord(word);
    }

    @Override
    protected void chosenCorrect(Button button) {
        setCorrect(button);
        setButtonsDisabled();
        usersBot.moveForward();
        if (usersBot.isWinner()) {
            applyWin();
            return;
        }
        button.postDelayed(new Runnable() {
            @Override
            public void run() {
                setButtonsEnabled();
                setNewRandomWord();
            }
        }, RACE_NEW_WORD_DELAY);
    }

    private void applyWin() {
        moveLogic.stopBots();
        concept.increaseLevel(this);
        showWinningDialog();
    }

    private void showWinningDialog() {
        final PromptDialog promptDialog = new PromptDialog(this);
        promptDialog
                .setDialogType(PromptDialog.DIALOG_TYPE_SUCCESS)
                .setAnimationEnable(false)
                .setTitleText(getString(R.string.congratulations))
                .setContentText(createDialogText())
                .setPositiveListener(getString(R.string.ok), new PromptDialog.OnPositiveListener() {
                    @Override
                    public void onClick(PromptDialog dialog) {
                        dialog.dismiss();
                        RaceActivity.this.finish();
                    }
                });
        promptDialog.setCanceledOnTouchOutside(false);
        promptDialog.show();
    }

    private String createDialogText() {
        StringBuilder s = new StringBuilder();
        if (concept.getCurrentLevel() <= concept.getNumberOfLevels()) {
            s.append("Jdete do levelu číslo ");
            s.append(concept.getCurrentLevel());
            s.append(" v kategorii ");
            s.append(concept.getName());
        } else {
            s.append("Dosáhli jste maximálního levelu v kategorii ");
            s.append(concept.getName());
        }
        s.append(".");
        return s.toString();
    }

    @Override
    protected void chosenWrong(Button button) {
        setWrong(button);
        usersBot.moveBackward();
    }

    @Override
    protected void onPause() {
        super.onPause();
        moveLogic.stopBots();
    }

    @Override
    protected void onStop() {
        super.onStop();
        this.finish();
    }
}

File: app/src/main/java/cz/muni/fi/umimecesky/game/robots/RaceActivity.kt
package cz.muni.fi.umimecesky.game.robots

import android.content.res.Resources
import android.os.Bundle
import android.widget.Button
import android.widget.ImageView
import cz.muni.fi.umimecesky.R
import cz.muni.fi.umimecesky.db.helper.joinCategoryWordOpenHelper
import cz.muni.fi.umimecesky.game.ball.Dimensions.isTablet
import cz.muni.fi.umimecesky.game.robots.RobotDrawable.getRobotDrawable
import cz.muni.fi.umimecesky.game.robots.logic.BotLogicQuick
import cz.muni.fi.umimecesky.game.robots.logic.BotLogicSlow
import cz.muni.fi.umimecesky.game.robots.logic.MoveLogic
import cz.muni.fi.umimecesky.game.shared.BaseAbstractActivity
import cz.muni.fi.umimecesky.game.shared.util.Constant.RACE_NEW_WORD_DELAY_MS
import cz.muni.fi.umimecesky.game.shared.util.Constant.RAW_HOPS_TO_WIN
import cz.muni.fi.umimecesky.prefs
import kotlinx.android.synthetic.main.activity_race.botView1
import kotlinx.android.synthetic.main.activity_race.botView2
import kotlinx.android.synthetic.main.activity_race.botView3
import kotlinx.android.synthetic.main.activity_race.finishLine
import kotlinx.android.synthetic.main.activity_race.firstButton
import kotlinx.android.synthetic.main.activity_race.secondButton
import kotlinx.android.synthetic.main.activity_race.usersRobotView
import kotlinx.android.synthetic.main.activity_race.word
import org.jetbrains.anko.configuration
import org.jetbrains.anko.landscape

// for RobotAnimator

class RaceActivity : BaseAbstractActivity() {

    private val concept = prefs.currentRobotConcept
    private val totalMovesToWin = RAW_HOPS_TO_WIN + concept.currentLevel

    private val screenWidthDp: Float
        get() = Resources.getSystem().displayMetrics.widthPixels.toFloat().dp

    private val robotWidthDp: Float
        get() = usersRobotView.layoutParams.width.toFloat().dp

    private lateinit var moveLogic: MoveLogic
    private lateinit var usersAnimator: RobotAnimator
    private var robotMovePx: Float = 10f

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_race)
        super.initUI(word, firstButton, secondButton)
        if (configuration.landscape && !isTablet()) supportActionBar?.hide()

        robotMovePx = ((screenWidthDp - robotWidthDp) / totalMovesToWin).px
        finishLine.x = (screenWidthDp - robotWidthDp - 2).px // 2 dp is line thickness
        setupRandomDrawables(arrayOf(botView1, botView2, botView3))

        usersAnimator = RobotAnimator(usersRobotView, Robot(), totalMovesToWin, robotMovePx)
        moveLogic = MoveLogic(this, createAnimators())

        setNewRandomWord()
    }


    private fun setupRandomDrawables(botViews: Array<ImageView>) {
        for (botView in botViews) {
            botView.setImageDrawable(getRobotDrawable(this))
        }
    }

    private fun createAnimators(): Array<RobotAnimator> {
        return arrayOf(
                usersAnimator,
                RobotAnimator(botView1, Robot(BotLogicQuick(concept)), totalMovesToWin, robotMovePx),
                RobotAnimator(botView2, Robot(BotLogicSlow(concept)), totalMovesToWin, robotMovePx),
                RobotAnimator(botView3, Robot(BotLogicQuick(concept)), totalMovesToWin, robotMovePx)
        )
    }

    private fun setNewRandomWord() =
            setWord(joinCategoryWordOpenHelper.getRandomCategoryWord(concept))

    override fun chosenCorrect(button: Button) {
        setCorrect(button)
        disableButtons()
        usersAnimator.applyCorrect()
        delayNewWord(button)
    }

    private fun delayNewWord(button: Button) {
        button.postDelayed({
            enableButtons()
            setNewRandomWord()
        }, RACE_NEW_WORD_DELAY_MS)
    }

    override fun chosenWrong(button: Button) {
        setWrong(button)
        usersAnimator.applyWrong()
    }

    override fun onPause() {
        super.onPause()
        moveLogic.stopBotsAndInput()
    }

    override fun onStop() {
        super.onStop()
        this.finish()
    }

}

val Float.dp: Float
    get() = this / Resources.getSystem().displayMetrics.density
val Float.px: Float
    get() = this * Resources.getSystem().displayMetrics.density